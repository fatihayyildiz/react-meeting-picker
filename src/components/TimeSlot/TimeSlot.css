.time-slot-box {
	padding: 5px;
	text-align: center;
	border: 1px solid #e1e1e1;
	border-radius: 2px;
	margin: 5px auto 0 !important;
	cursor: pointer;
	background-color: white;
}

.disabled {
	background-color: rgba(209, 209, 209, 0.65);
	cursor: not-allowed;
}

.selected {
	background-color: rgba(111, 99, 229, 0.75);
}
