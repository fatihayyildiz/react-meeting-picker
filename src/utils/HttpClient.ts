// Generic Axios instance, heavily inspired via this post:
// https://levelup.gitconnected.com/enhance-your-http-request-with-axios-and-typescript-f52a6c6c2c8e
import axios, {
	AxiosError,
	AxiosInstance,
	AxiosRequestHeaders,
	AxiosResponse,
} from 'axios';

declare module 'axios' {
	// eslint-disable-next-line
	interface AxiosResponse<T = any> extends Promise<T> {}
}

// Abstract of axios instance
// Only handles response interceptor in abstract class
// Request interceptor should handle in inherited class
export abstract class HttpClient {
	protected readonly instance: AxiosInstance;

	protected constructor(
		baseURL: string,
		headers: AxiosRequestHeaders = {
			'Content-Type': 'application/json',
		},
		timeout: number = 30000
	) {
		this.instance = axios.create({
			baseURL,
			headers,
			timeout,
		});

		// Runs after every response from call
		this._initializeResponseInterceptor();
	}

	// Handles two case in below:
	// _handleResponse : Successful response from call
	// _handleError: Error case of call
	private _initializeResponseInterceptor = () => {
		this.instance.interceptors.response.use(
			this._handleResponse,
			this._handleError
		);
	};

	// Pass response object to Promise resolve
	private _handleResponse = ({ data }: AxiosResponse) => data;

	// Catch error and throw code to .catch block
	protected _handleError = (error: AxiosError) => Promise.reject(error);
}
