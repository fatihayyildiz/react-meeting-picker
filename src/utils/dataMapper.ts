import {
	APICompany,
	CompanySlots,
	DaySlots,
	TimeSlot,
} from 'store/slots/types';
import dayjs from 'dayjs';
import LocalizedFormat from 'dayjs/plugin/localizedFormat';

dayjs.extend(LocalizedFormat);

export const convertApiResponseToFEModel = (companies: Array<APICompany>) => {
	const convertedCompanySlots = companies.map((company: APICompany) => {
		const daysArray = company.time_slots.reduce(
			(acc: Array<DaySlots>, curr) => {
				let currentDay = dayjs(curr.start_time).format('LL');
				let foundIndexInDays = acc.findIndex((day) => day.day === currentDay);

				if (foundIndexInDays > -1) {
					const foundedCurrentDay = acc[foundIndexInDays] as DaySlots;
					foundedCurrentDay.slots.push({
						day: currentDay,
						start_time: dayjs(curr.start_time).format('HH:mm'),
						end_time: dayjs(curr.end_time).format('HH:mm'),
					});
				} else {
					const newDaySlot = {
						day: currentDay,
						start_time: dayjs(curr.start_time).format('HH:mm'),
						end_time: dayjs(curr.end_time).format('HH:mm'),
					} as TimeSlot;
					acc.push({
						day: currentDay,
						slots: [newDaySlot],
					});
				}
				return acc;
			},
			[]
		);
		//console.log('converted days:',daysArray);
		return {
			id: company.id,
			name: company.name,
			days: daysArray as Array<DaySlots>,
		};
	}) as Array<CompanySlots>;

	console.log('converted companies:', convertedCompanySlots);
	return convertedCompanySlots;
};
