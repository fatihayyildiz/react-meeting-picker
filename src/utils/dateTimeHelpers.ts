import dayjs from 'dayjs';
import isBetween from 'dayjs/plugin/isBetween';
import { TimeSlot } from 'store/slots/types';

dayjs.extend(isBetween);

export const checkDateIsBetween = ({
	selectedSlot,
	day,
	start_time,
	end_time,
}: {
	selectedSlot: TimeSlot;
	day: string;
	start_time: string;
	end_time: string;
}): boolean => {
	// Logic for detect selected time slot is between other time slots
	// Used for disable other time slots if agent occupied with these times
	// add and removes one minute to calculate correct range
	return (
		(dayjs(
			dayjs(`${dayjs(day).format('YYYY-MM-DD')} ${start_time}`).add(1, 'minute')
		).isBetween(
			`${dayjs(day).format('YYYY-MM-DD')} ${selectedSlot.start_time}`,
			dayjs(
				`${dayjs(day).format('YYYY-MM-DD')} ${selectedSlot.end_time}`
			).subtract(1, 'minute'),
			'minutes',
			'[]'
		) ||
			dayjs(
				dayjs(`${dayjs(day).format('YYYY-MM-DD')} ${end_time}`).subtract(
					1,
					'minute'
				)
			).isBetween(
				`${dayjs(day).format('YYYY-MM-DD')} ${selectedSlot.start_time}`,
				dayjs(
					`${dayjs(day).format('YYYY-MM-DD')} ${selectedSlot.end_time}`
				).subtract(1, 'minute'),
				'minutes',
				'[]'
			)) &&
		dayjs(dayjs(selectedSlot.day).format('YYYY-MM-DD')).isBetween(
			dayjs(dayjs(day).format('YYYY-MM-DD')).subtract(1, 'day'),
			dayjs(dayjs(day).format('YYYY-MM-DD')).add(1, 'day'),
			'day'
		)
	);
};
