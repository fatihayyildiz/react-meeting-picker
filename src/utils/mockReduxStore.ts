import { CompanySelectedSlot, CompanySlots } from 'store/slots/types';

export const MockReduxStoreInitialState = {
	slots: {
		loading: false,
		companies: [] as Array<CompanySlots>,
		selectedCompanySlots: [] as Array<CompanySelectedSlot>,
	},
};
