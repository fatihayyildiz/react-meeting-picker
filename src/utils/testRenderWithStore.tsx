import React from 'react';
import { configureStore } from '@reduxjs/toolkit';
import { Provider } from 'react-redux';
import { render as renderRtl } from '@testing-library/react';
import { RootState } from 'store/store';
import { reducers } from 'store/reducers';

interface WrapperProps {
	children?: React.ReactNode;
}

const testStore = (state: Partial<RootState>) => {
	return configureStore({
		reducer: reducers,
		preloadedState: state,
	});
};

export const testRenderWithStore = (
	component: React.ReactNode,
	initialState: any
) => {
	const wrapper = ({ children }: WrapperProps) => (
		<Provider store={testStore(initialState)}>{children}</Provider>
	);
	return renderRtl(component as any, { wrapper });
};
