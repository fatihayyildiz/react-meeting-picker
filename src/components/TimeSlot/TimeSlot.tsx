import React from 'react';
import 'components/TimeSlot/TimeSlot.css';
import classNames from 'classnames';
import { useDispatch } from 'react-redux';
import { AppDispatch } from 'store/store';
import slotSlice from 'store/slots/slice';

export type TimeSlotProps = {
	id: number;
	day: string;
	start_time: string;
	end_time: string;
	available?: boolean;
	selected?: boolean;
};

const TimeSlot = ({
	id,
	day,
	start_time,
	end_time,
	available = true,
	selected = false,
}: TimeSlotProps) => {
	const dispatch = useDispatch<AppDispatch>();

	const onSlotClick = () => {
		dispatch(
			slotSlice.actions.selectSlotForCompany({ id, day, start_time, end_time })
		);
	};

	return (
		<div
			className={classNames('row', 'time-slot-box', {
				available: available,
				selected: selected,
			})}
			onClick={() => onSlotClick()}
		>
			<div className="col-4">{start_time}</div>
			<div className="col-2">-</div>
			<div className="col-4">{end_time}</div>
		</div>
	);
};

export default TimeSlot;
