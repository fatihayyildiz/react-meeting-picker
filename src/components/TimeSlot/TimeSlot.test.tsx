import React from 'react';
import { cleanup, screen, RenderResult } from '@testing-library/react';
import TimeSlot from 'components/TimeSlot/TimeSlot';
import { testRenderWithStore } from 'utils/testRenderWithStore';
import { MockReduxStoreInitialState } from 'utils/mockReduxStore';

afterEach(cleanup);

const testProps = {
	id: 1,
	day: 'Monday July 9, 2018',
	start_time: '08:00',
	end_time: '09:00',
	available: true,
	selected: true,
};

const renderTimeSlot = (): RenderResult =>
	testRenderWithStore(
		<TimeSlot
			end_time={testProps.end_time}
			start_time={testProps.start_time}
			id={testProps.id}
			day={testProps.day}
		/>,
		MockReduxStoreInitialState
	);

test('Timeslot should match with snapshot', () => {
	// First render component and get as fragment
	const { asFragment } = renderTimeSlot();

	// Compare with saved snapshot
	expect(asFragment()).toMatchSnapshot();
});

test('Timeslot renders times correctly', () => {
	// First render component with props and get as fragment
	renderTimeSlot();

	// Compare start time
	expect(screen.getByText(testProps.start_time)).toHaveTextContent(
		testProps.start_time
	);

	// Compare end time
	expect(screen.getByText(testProps.end_time)).toHaveTextContent(
		testProps.end_time
	);
});

test('Timeslot renders conditional classnames correctly', () => {
	// First render component with props and get as fragment
	const { container } = renderTimeSlot();

	// eslint-disable-next-line testing-library/no-container, testing-library/no-node-access
	const disabled = container.getElementsByClassName('disabled');

	// eslint-disable-next-line testing-library/no-container, testing-library/no-node-access
	const selected = container.getElementsByClassName('selected');

	// Check disabled slot
	expect(disabled.length).toBe(0);

	// Check selected class
	expect(selected.length).toBe(0);
});
