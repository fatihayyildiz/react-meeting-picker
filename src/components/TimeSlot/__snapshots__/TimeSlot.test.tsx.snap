// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Timeslot should match with snapshot 1`] = `
<DocumentFragment>
  <div
    class="row time-slot-box"
  >
    <div
      class="col-4"
    >
      08:00
    </div>
    <div
      class="col-2"
    >
      -
    </div>
    <div
      class="col-4"
    >
      09:00
    </div>
  </div>
</DocumentFragment>
`;
